File: problems/leetcode/1448.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
    
        def checksubtree(root, greatestAbove):
            if not root:
                return 0
            res = 1 if root.val >= greatestAbove else 0
            greatestAbove = max(greatestAbove, root.val)
            res += checksubtree(root.left, greatestAbove)
            res += checksubtree(root.right, greatestAbove)
            return res

        return checksubtree(root, root.val)

File: problems/leetcode/1448.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        return checkSubtree(root, root->val);
    }
    
    int checkSubtree(TreeNode* root, int greatestAbove) {
        if (!root) {
            return 0;
        }
        int res = root->val>=greatestAbove ? 1 : 0;
        greatestAbove = max(greatestAbove, root->val);
        res += checkSubtree(root->left, greatestAbove);
        res += checkSubtree(root->right, greatestAbove);
        return res;
    }
};